framework/src/csg/CSGRegion.C

```diff
@@ -0,0 +1,189 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CSGRegion.h"
+
+namespace CSG
+{
+
+CSGRegion::CSGRegion()
+{
+  _region_str = "";
+  _region_type = "EMPTY";
+}
+
+// halfspace constructor
+CSGRegion::CSGRegion(const CSGSurface & surf, const CSGSurface::Halfspace halfspace)
+{
+  _region_type = "HALFSPACE";
+  _region_str = ((halfspace == CSGSurface::Halfspace::POSITIVE) ? "+" : "-") + surf.getName();
+  _surfaces.push_back(surf);
+}
+
+// intersection and union constructor
+CSGRegion::CSGRegion(const CSGRegion & region_a,
+                     const CSGRegion & region_b,
+                     std::string region_type)
+{
+  _region_type = region_type;
+  if (getRegionType() != RegionType::INTERSECTION && getRegionType() != RegionType::UNION)
+    mooseError("Region type " + getRegionTypeString() + " is not supported for two regions.");
+  else if (region_a.getRegionType() == RegionType::EMPTY ||
+           region_b.getRegionType() == RegionType::EMPTY)
+    mooseError("Region operation " + getRegionTypeString() +
+               " cannot be performed on an empty region.");
+  else
+  {
+    std::string op = (getRegionType() == RegionType::UNION) ? " | " : " & ";
+    auto a_string = stripRegionString(region_a.toString(), op);
+    auto b_string = stripRegionString(region_b.toString(), op);
+
+    _region_str = "(" + a_string + op + b_string + ")";
+    const auto & a_surfs = region_a.getSurfaces();
+    const auto & b_surfs = region_b.getSurfaces();
+    _surfaces.insert(_surfaces.end(), a_surfs.begin(), a_surfs.end());
+    _surfaces.insert(_surfaces.end(), b_surfs.begin(), b_surfs.end());
+  }
```

Comment:
```suggestion
  if (getRegionType() != RegionType::INTERSECTION && getRegionType() != RegionType::UNION)
    mooseError("Region type " + getRegionTypeString() + " is not supported for two regions.");
  if (region_a.getRegionType() == RegionType::EMPTY ||
           region_b.getRegionType() == RegionType::EMPTY)
    mooseError("Region operation " + getRegionTypeString() +
               " cannot be performed on an empty region.");

  std::string op = (getRegionType() == RegionType::UNION) ? " | " : " & ";
  auto a_string = stripRegionString(region_a.toString(), op);
  auto b_string = stripRegionString(region_b.toString(), op);

  _region_str = "(" + a_string + op + b_string + ")";
  const auto & a_surfs = region_a.getSurfaces();
  const auto & b_surfs = region_b.getSurfaces();
  _surfaces.insert(_surfaces.end(), a_surfs.begin(), a_surfs.end());
  _surfaces.insert(_surfaces.end(), b_surfs.begin(), b_surfs.end());
```

Minor - but a bit more readable